### ISSUE TITLE
Calculate frame luminance using histogram for HDR

### ISSUE THREAD

**Issue by mrDIMAS** [FyroxEngine/Fyrox#587]
Currently, frame luminance is calculated using downsampling with bilinear filtration, that basically calculates _average_ frame luminance. Average luminance works quite bad in HDR, because a single bright pixel could easily ruin the luminance values. It could be done either on CPU or using a compute shader (this must be gated by checking if compute shaders are actually available).

**Comment by The-Ladle** [comment]
I want to try and take this issue, but I have no idea where frame luminance is calculated in the engine. If you could let me know, I'd be happy to start and try to fix this.

**Comment by mrDIMAS** [comment]
It is in `HighDynamicRangeRenderer::calculate_avg_frame_luminance` and I already did some basic version of histogram-based luminance calculation, but it requires tweaking and probably contains bugs in its implementation.

**Comment by The-Ladle** [comment]
I have been unable to figure this one out, which means I've effectively given up on it.

**Comment by EricMiddelhove** [comment]
I would like to check it out, I cant guarantee that I will be able to solve it. But I'll try!

**Comment by EricMiddelhove** [comment]
I would appreciate a clarification on what is expected of the luminance calculation. I believe I understand the calculation happening in the current implementation however, a clarification on the use case of the calculation would also provide insight on how to test the calculation. @mrDIMAS

**Comment by EricMiddelhove** [comment]
https://github.com/EricMiddelhove/Fyrox/commit/9a4a9a4178110fd3d4847b0d30345f04272f4970

In this commit in my Fork of the project I suggest a refactoring, attempting to simplify the calculation with the same effect. Please let me know if this is what you were thinking, or what I should be changed.

**Comment by mrDIMAS** [comment]
If I recall correctly, my attempt of histogram calculation was unsuccessful, because I simply got confused by the math and just skipped it. It may not even work correctly, I can't really remember.

The desired effect is to prevent the screen from darkening when there's just one or two very bright pixels. Histogram allows you to filter out such pixels and only then calculate the luminance. So it is probably worth creating a scene with some bright objects of different size, that have large emission values on their PBR material and fly around with the camera to see if adaptation work correctly. Dumb approach with simply averaging all the pixels works poorly on such scene, a singe small but bright object makes the screen dark because its luminance contribution is very high. Histogram approach should discard some of the very bright and maybe very dim pixels and only then calculate the luminance. 

To simply the testing, you can turn on the HDR in the editor by changing editor camera's exposure value to `Exposure::Auto` (https://github.com/FyroxEngine/Fyrox/blob/master/editor/src/camera/mod.rs#L155) with default's that can be seen in the docs for `Exposure::Auto`.

**Comment by EricMiddelhove** [comment]
Okay! Perfect! I will test my code adjustments!

**Comment by EricMiddelhove** [comment]
I am currently not able to reproduce the issue. Can you provide me with a scene or project where this issue occurrs?

**Comment by mrDIMAS** [comment]
I'll make a test scene closer to the end of the day to help you to reproduce this issue. Probably submit it on a separate branch.

**Comment by mrDIMAS** [comment]
@EricMiddelhove I've submitted changes on `hdr_test` branch, it contains `unnamed.rgs` scene which should help you to understand the issues with HDR in the engine in general. At first, note how dim the picture is with turned on HDR and it only gets lighter when you look directly into the black ground. Next, when you look at the bright cubes, the picture gets even dimmer. 

So I think HDR in the engine is quite broken and if you could fix it, that would be super awesome. There are number of papers online that explains how to tackle this problem in general. If you have any questions, just ask and I'll help.

**Comment by EricMiddelhove** [comment]
Thats awesome! Thanks! I will do my best fixing it :)

**Comment by EricMiddelhove** [comment]
![Image](https://github.com/user-attachments/assets/977324db-ff62-4822-b5fd-4ef57fc2df47)

![Image](https://github.com/user-attachments/assets/a5679d87-50f4-4423-b784-4ccae4fa56f8)

Does this look better?
The first is the original the second is the fix

**Comment by EricMiddelhove** [comment]
![Image](https://github.com/user-attachments/assets/d3c58fd8-a1ca-433f-a806-1de1326a8a1f)

![Image](https://github.com/user-attachments/assets/577e5049-f16e-48ed-8a5d-736a08a0cd00)

This is the situation with the sponza scene @mrDIMAS

**Comment by mrDIMAS** [comment]
That looks good, how does it look in dynamic? 

As for `log2` - I've read this article - https://bruop.github.io/exposure/ and can't really remember why `log2` is used in there to be honest.

**Comment by EricMiddelhove** [comment]
https://youtu.be/d9WDwndNaQo

@mrDIMAS I uploaded a video of the fix attempt to youtube

**Comment by mrDIMAS** [comment]
Ok, that looks nice, what's the light intensity on the video?

**Comment by EricMiddelhove** [comment]
The intensity is 1. 

In my most recent commit to my fork, I adjusted the color range which should make the exposure adjustment more smoothly. If you test it and sign it off, I would refactor it to make it more maintainable and then I would prepare a PR. :)

**Comment by mrDIMAS** [comment]
Yeah, I'd like to see a PR. Also, I think the overall intensity thingy should somehow be based off physical values, but that's for future work I suppose.

**Comment by EricMiddelhove** [comment]
Is this issue considered closed, or will it change? :)
